lib/camp.js

```diff
@@ -628,13 +635,14 @@ function respondWithFile(req, res, path, ifNoFile) {
     }
     res.mime(p.extname(realpath).slice(1));
 
-    // Cache management.
-    if (+(new Date(req.headers['if-modified-since'])) >= +stats.mtime) {
+    // Cache management (compare timestamps with second-level precision).
+    var since = req.headers['if-modified-since'];
+    if (since && (Math.abs(stats.mtime - new Date(since)) < 1000)) {
```

Comment:
If I understand correctly, browsers truncate to the start of the second, so this doesn't actually work as intended either.

For instance:
- a file is changed at xxxx-xx-xxTxx-xx-01.02Z
- it is requested, producing a registered client-side last modification time of xxxx-xx-xxTxx-xx-01Z
- the file is changed at xxxx-xx-xxTxx-xx-01.98Z
- a long time later, the file is requested with an if-modified-since header.

Since the if-modified-since and the mtime are less than 1000 milliseconds apart (they are 996 ms apart), we send a 304 even though the file has changed — *and the change will never be delivered* until the cache is cleared.

One way to see this is that your code does this:

                       update
    ----1s----|----1s----|----1s----|----1s----
               xxxxxxxxxxxxxxxxxxxxx
              zone in which previous updates are assumed to be that update

You'll notice it is also vulnerable to leap seconds. If an update occurs at the beginning of the first leap second, and another update occurs in the middle of the second pass over the leap second (ie, 1.5s later), future requests will assume this is the same update (and send a 304, therefore never delivering the second update) even though the updates are more than 1s apart.

I understand that this issue limits the usefulness of the If-Modified-Since header. The solution is to make use of the ETag / If-None-Match headers. You may put the millisecond mtime in there… until we finally get nanosecond precision for mtime, reducing possible collisions to a much more acceptable level.